File: python/epi/reverse_sublist.py
#!/usr/bin/env python3
#
# Description:
#
# Reverse a sublist of a linked-list given the head and a zero-based, inclusive,
# start and end integer boundary.

from rlib.linked_list_node import LinkedListNode
from rlib.list_utils import buildSingleList
from rlib.list_utils import draw
from rlib.testing_utils import expect

def reverseSublist(head, start, end):
    if end - start < 1 or head is None: return head
    dummy = LinkedListNode(0, head)
    prv = dummy
    cur = head
    counter = 0
    while counter < start:
        cur = cur.next
        prv = prv.next
        counter += 1
    subp = prv
    subh = cur
    nxt = cur.next
    while counter < end and nxt is not None:
        tmp = nxt.next
        nxt.next = cur
        subp.next = nxt
        subh.next = tmp
        cur = nxt
        nxt = tmp
        counter += 1
    return dummy.next

def check(headA, headB):
    if headA is None and headB is None: return True
    while headA and headB:
        if headA.data != headB.data: return False
        headA, headB = headA.next, headB.next
    if headA is None and headB is None: return True
    else: return False

def small_test():
    head = buildSingleList([1, 2, 3, 4, 5])
    result = reverseSublist(head, 0, 0)
    expected = buildSingleList([1, 2, 3, 4, 5])
    expect(check(result, expected)).to_be(True)
    head = buildSingleList([1, 2, 3, 4, 5])
    result = reverseSublist(head, 1, 3)
    expected = buildSingleList([1, 4, 3, 2, 5])
    expect(check(result, expected)).to_be(True)
    head = buildSingleList([1, 2, 3, 4, 5])
    result = reverseSublist(head, 0, 4)
    expected = buildSingleList([5, 4, 3, 2, 1])
    expect(check(result, expected)).to_be(True)

if __name__ == '__main__':
    small_test()

File: java/epi_old/ReverseSublist.java
// Description:
//
// Reverse a sublist of a linked-list given a starting and ending position, inclusive.

package epi;

import java.util.Arrays;
import java.util.List;
import rlib.ListUtils;
import rlib.LinkedListNode;
import static rlib.TestingUtils.expect;

public class ReverseSublist {

  public static LinkedListNode<Integer> reverseSublist(
    LinkedListNode<Integer> head, int start, int end)
  {
    LinkedListNode<Integer> dummy = new LinkedListNode<>(null, head);
    LinkedListNode<Integer> subp = dummy;
    int counter = 0;
    while (counter < start) {
      subp = subp.next;
      counter++;
    }
    LinkedListNode<Integer> curr = subp.next;
    while (start < end) {
      LinkedListNode<Integer> temp = curr.next;
      curr.next = temp.next;
      temp.next = subp.next;
      subp.next = temp;
      start++;
    }
    return dummy.next;
  }

  private static boolean check(
    LinkedListNode<Integer> head1, LinkedListNode<Integer> head2)
  {
    while (head1 != null || head2 != null) {
      if (head1.data != head2.data) return false;
      head1 = head1.next;
      head2 = head2.next;
    }
    if (head1 == null && head2 == null) return true;
    else return false;
  }

  private static void smallTest() {
    LinkedListNode<Integer> head, result, expected;

    head = ListUtils.buildSingleList(Arrays.asList(1, 2, 3, 4, 5, 6, 7, 8, 9, 10));
    result = reverseSublist(head, 1, 8);
    expected = ListUtils.buildSingleList(Arrays.asList(1, 9, 8, 7, 6, 5, 4, 3, 2, 10));
    expect(check(result, expected)).toBe(true);

    head = ListUtils.buildSingleList(Arrays.asList(1));
    result = reverseSublist(head, 0, 0);
    expected = ListUtils.buildSingleList(Arrays.asList(1));
    expect(check(result, expected)).toBe(true);

    head = ListUtils.buildSingleList(Arrays.asList(1, 2));
    result = reverseSublist(head, 0, 1);
    expected = ListUtils.buildSingleList(Arrays.asList(2, 1));
    expect(check(result, expected)).toBe(true);

    head = ListUtils.buildSingleList(Arrays.asList(1, 2));
    result = reverseSublist(head, 1, 1);
    expected = ListUtils.buildSingleList(Arrays.asList(1, 2));
    expect(check(result, expected)).toBe(true);
  }

  public static void main(String[] args) {
    smallTest();
  }

}
